Windows Cmdlets for CI/CD (#5047)
mendix/docs: content/en/docs/developerportal/deploy/on-premises-design/ms-windows/automate-mendix-deployment-on-microsoft-windows.md

---
title: "Automate Mendix Deployment on Microsoft Windows"
url: /developerportal/deploy/automate-mendix-deployment-on-microsoft-windows/
description: "How to automate Mendix deployment on servers running Windows"
weight: 5
tags: ["deploy", "Windows", "Mendix Service Console", "Microsoft", "CI/CD pipeline", "cmdlets"]
---

## 1 Introduction

> On Windows servers, instead of deploying your application manually, you can automate that part of your CI/CD pipeline by using [Power Shell cmdlets](https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/cmdlet-overview). This document describes the required configuration, and provides sample automation scripts that you can use as basis for scripting your own deployment.

Author: I would like to expand this section with more info about when you would or wouldn't want to automate deployment. I'd expect that most clients would want to avoid manual deployment, but apparently this isn't the case, and I'm wondering why.

Reviewer: Perhaps you can focus on the benefits of automating a deployment. There isn't really a good reason _not_ to automate, it's just the time and effort that would go into setting it up.

Author: Added a sentence outlining the benefits. @tieniber would you say that Mendix recommends automating the deployment? Currently this document is separated from the one about manual deployments, but if automation is the happy path then maybe we should combine them.

Reviewer 2: @katarzyna-koltun-mx 
Mendix was initially based around developers who would make apps and release them quickly. There are functions to push to the cloud directly from within Studio and Studio Pro, and simple processes for transporting from development to production within the Developer Portal.
When it comes to other platforms, things become more difficult, but the concept is still that the developer publishes when they are ready to do so.
It is only when customers are larger and have more automated pipelines (and automated tests, in particular) that CI/CD becomes worthwhile. However, it is not necessarily "better" than going through the manual process, or in any way the "happy path".

- CI/CD benefit - developer doesn't have to worry about initiating a build, can add automated testing to pipelines
- Manual benefit - can choose when to build, can control when app downtime occurs (and use a human decision, rather than pre-built rules).

I would say they are two options, but that the manual version is easier to understand and easier to use out of the box. So I would suggest keeping this as an additional document.

However, @tieniber works more with customers directly, so he may have other insights.

Author: @MarkvanMents that makes sense - do you think I should add that info to the doc? Especially the pros and cons of both would be helpful. Some clients will definitely know one way or the other, but others might not be sure which option to choose, and a comparison like what you outlined above would be helpful for them to make the decision.

Author: Discussed offline, decided to add it to https://docs.mendix.com/developerportal/deploy/general/ as it outlines some general concerns, not just stuff that's specific to Windows

## 2 Prerequisites

> Before starting this how-to, make sure you have completed the following prerequisites:

Reviewer 2: I think the tense should match with the prerequisites. So I would either change the three bullet points below into the past, or say:
```suggestion
Before starting this how-to, make sure you complete the following prerequisites:
```

## 3 Importing Mendix-specific Cmdlets into Windows PowerShell

To install Mendix-specific cmdlets that you can use to script your app deployment, follow these steps:

1. Build a Mendix Deployment Package by choosing one of the following options:
    * Use the Mendix Build API. For more information, see [Build API](/apidocs-mxsdk/apidocs/build-api/).
    * Fetch the source from Team server and build the package locally with [MxBuild.exe](/refguide/mxbuild/).
    * Create the package manually. For more information, see [Create Deployment Package](/refguide/create-deployment-package-dialog/).
2. In Windows PowerShell, run the following command: `Import-Module '{<Mendix Service Console installation directory>}\Mendix.Service.Commands.dll'`.
    For example, if Mendix Service Console is installed at *C:\Program Files (x86)\Mendix\Service Console*, enter `Import-Module 'C:\Program Files (x86)\Mendix\Service Console\Mendix.Service.Commands.dll'`
3. Verify that the following commands are now available in PowerShell:
    * `Start-MxApp`
    * `Stop-MxApp`
    * `Update-MxApp`
    * `Install-MxServer`

>     * `Get-MxApp`

Reviewer: We never provide a sample of Get-MxApp later in the doc. Should we? I'm actually not sure what value it provides.

Author: I'm not sure myself - @MarkvanMents who would be the best person to ask?

Reviewer 2: Well, if @tieniber doesn't think it provides any additional value, then we should consider removing it. The two people I think who might be able to help are @jveldhuizen or @feddovanede. You could try asking them on Slack (or here!). If they think it is useful, then we should give an example.

Reviewer 3: At this moment, it's not so useful, so it's fine to not mention the `Get-MxApp` command. If you want to keep it, an example could be

`Get-MxApp MyFirstApp | Start-MxApp -synchronizedatabase`

because the output of Get-MxApp is an App object that you can forward to the next command.
Another forwarding example - not related to Get-MxApp - is:

`dir '.\deployment-file.mda' | Update-MxApp -SynchronizeDatabase | Start-MxApp`

Author: I've decided to remove the command since it's not so useful at this time.

    
    For more information about each command and its parameters, add `-?` after the command, for example, `Start-MxApp -?`.

After installing the Mendix-specific cmdlets, you can use them to write your own scripts that start, stop, or update your Mendix app.

## 4 Sample Scripts

In this section, you can find sample scripts to help you script your app deployment.

{{% alert color="info" %}}
The scripts are intended to show the range of available deployment options. They are presented as examples only, and may require significant adaptation to work in your own environment.
{{% /alert %}}

### 4.1 Sample Script - Update App

> The following script example demonstrates the process required to update your app. First, it imports the necessary cmdlets. After that, it stops your app and updates it with files extracted from the Mendix Deployment Package. Finally, the script restarts the app with the `SynchronizeDatabase` parameter to synchronize the database without user input. Without this parameter, you must give additional confirmation to execute the database commands.

Reviewer 2: I think "Firstly" to match "Finally"?
```suggestion
The following script example demonstrates the process required to update your app. Firstly, it imports the necessary cmdlets. After that, it stops your app and updates it with files extracted from the Mendix Deployment Package. Finally, the script restarts the app with the `SynchronizeDatabase` parameter to synchronize the database without user input. Without this parameter, you must give additional confirmation to execute the database commands.
```

```text {linenos=table}
Import-Module '{<Mendix Service Console installation directory>}\Mendix.Service.Commands.dll'                                                                                                                        
$MDA_PATH = Read-Host '{Location of your Mendix Deployment Package}'       
$APP_NAME = Read-Host '{Name of your app}'         

"Deploying " + $MDA_PATH + " to app " + $APP_NAME                         

# stop app                                                      
Stop-MxApp $APP_NAME                                                    

# unpack app                                                    
Update-MxApp $APP_NAME -LiteralPath $MDA_PATH + "/" + $MDA_FILE

# start app, update database                                     
Start-MxApp MyFirstApp -SynchronizeDatabase
```

{{% alert color="warning" %}}

> Stopping your app before you update it is a necessary part of the process. Do not attempt to extract the deployment package into your app while the app is running.

Reviewer 2: Will this just fail, or is there a risk that it will break something?
Hopefully, the former. In which case we can say "if you attempt to extract … your script will fail". Which is a bit less scary!

Reviewer 3: Sorry, this makes your app corrupt and it cannot be restarted anymore...

If you do this, you get the error (sorry for the Dutch texts):
```
WARNING: An error occurred while deleting folder 'C:\Mendix\Uitrolling\Apps\Ticket125615\Project\model\bundles'. Specific error: Kan geen toegang krijgen tot het bestand C:\Mendix\Uitrolling\Apps\Ticket125615\Project\model\bundles\project.jar omdat het wordt gebruikt
door een ander proces. (C:\Mendix\Uitrolling\Apps\Ticket125615\Project\model\bundles\project.jar)
De map is niet leeg.
 (C:\Mendix\Uitrolling\Apps\Ticket125615\Project\model\bundles)

The update will be proceeded successfully, regardless this error.
Update-MxApp : Het proces kan het bestand niet openen omdat het door een ander proces wordt gebruikt.
At line:1 char:86
+ ... 5\Project\releases\App_20210803_1602.mda" | Update-MxApp Ticket125615
+                                                 ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (Ticket125615:String) [Update-MxApp], IOException
    + FullyQualifiedErrorId : AppProcessError,Mendix.Service.Commands.UpdateAppCommand
```
It says

> The update will be proceeded successfully, regardless this error.

but this is actually not true!
If you stop and then start the app, you get the error
`Start-MxApp : Service cannot be started. See log messages for details.`
and the Windows Event Viewer shows a shocking server stack trace.

OK, it's not so bad, because if you - while your app is stopped - update it again, it will be succesful and you can start successfully again. But this is not the flow you want.

Author: So it seems that the heavy-sounding warning is appropriate for this - I'll keep it as is.

{{% /alert %}}

### 4.2 Sample Script - Update the Mendix Runtime

The following script example demonstrates how you can update the [Mendix Runtime](/refguide/runtime/) to a version that matches the app that you are deploying. This is only required when you upgrade Mendix versions. The sample script first downloads the required Mendix Runtime version through PowerShell, and then extracts the Mendix Platform libraries into the server distribution folder. In this case, the app can remain running, as this process only extracts the new server version without affecting previously installed Mendix Platform versions.

```text {linenos=table}
# download Mendix Runtime
wget https://cdn.mendix.com/runtime/mendix-{<major>.<minor>.<patch>.<build>}.tar.gz -OutFile {<target folder for the downloaded file>}\mendix-{<major>.<minor>.<patch>.<build>}.gz

# extract Mendix Platform into the distribution folder
Install-MxServer -LiteralPath {<target folder for the downloaded file>}\mendix-{<major>.<minor>.<patch>.<build>}.gz
```

> ### 4.3 Sample Script - Determine the Mendix Runtime Version

Reviewer: You might swap 4.2 and 4.3, since you'll want to know the Mx version you need before installing it

Author: Done.

The following script demonstrates how you can check which version of the Mendix Runtime is required for your app. It inspects a deployment package, finds the Mendix Runtime version it needs, and downloads the correct version.

```text {linenos=table}
Copy-Item -Path 'C:\Mendix\Some_Deployment.mda'-Destination C:\Temp\temp.zip
Expand-Archive -LiteralPath 'C:\Temp\temp.zip' -DestinationPath C:\temp\MxApp
$mxJson = Get-Content "C:\temp\MxApp\model\metadata.json" | ConvertFrom-Json
Remove-Item C:\Temp\temp.zip
Remove-Item C:\Temp\MxApp\ -Recurse

# determine the Mendix Runtime version
$mxJson.RuntimeVersion

# download the Mendix Runtime version
$targetURL = 'https://cdn.mendix.com/runtime/mendix-' + $mxJson.RuntimeVersion + ".tar.gz"
$targetFile = 'C:\Mendix\runtimes\mendix-' + $mxJson.RuntimeVersion + '.tar.gz'
wget $targetURL -OutFile $targetFile
```

> ## 5 Troubleshooting

Reviewer: There's one more error that can happen often. If you run a script without admin rights, you'll get a cryptic error message. I don't have exact details but did run into it when testing myself a few weeks ago. Recommend we add a section.

Author: Will do, can you give me more details about the error though? Is there an actual error message? Does the script just stop?

Reviewer 3: If the app's name is 'App1', the error message is:

```
Start-MxApp : Cannot open App1 service on computer '.'.
At line:1 char:1
+ Start-MxApp App1 -synchronizedatabase
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (App1:String) [Start-MxApp], InvalidOperationException
    + FullyQualifiedErrorId : AppProcessError,Mendix.Service.Commands.StartAppCommand
```

Author: @jveldhuizen Thank you! I've added a section to the end of the document. Let me know if anything needs correcting.

> If you encounter any issues while configuring automated Mendix deployment on Windows, see the following troubleshooting tips to help you solve them.

Reviewer 2: We have never really called it "automated Mendix deployment on Windows". Perhaps you can get around this by saying something like:
```suggestion
If you encounter any issues while automating Mendix deployment on Windows using cmdlets, use the following troubleshooting tips to help you solve them.
```

### 5.1 Could Not Load File or Assembly

PowerShell shows an error message similar to the following:

```text {linenos=false}
Could not load file or assembly, Version=3.3.0.0, Culture=neutral, PublicKeyToken= {token number}' or one of its dependencies. The system cannot find the file specified.
```

#### 5.1.1 Cause

This error or any similar error message is  likely related to a mismatch in the .NET version between PowerShell and the Mendix cmdlets. The Mendix cmdlets require .NET version 4, so the error may show if PowerShell is using .NET 2 or 3.

> #### 5.1.2 Solution

Reviewer: There's one more error that can happen often. If you run a script without admin rights, you'll get a cryptic error message. I don't have exact details but did run into it when testing myself a few weeks ago. Recommend we add a section.

> This error or any similar error message is  likely related to a mismatch in the .NET version between PowerShell and the Mendix cmdlets. The Mendix cmdlets require .NET version 4, so the error may show if PowerShell is using .NET 2 or 3.

Reviewer 2: Although there isn't anything wrong with what you have written, it feels better (more formal?) to me to write:
```suggestion
This error, or any similar error message, is likely to be related to a mismatch in the .NET version between PowerShell and the Mendix cmdlets. The Mendix cmdlets require .NET version 4, so the error may show if PowerShell is using .NET 2 or 3.
```

To solve this issue, follow these steps:

1. Determine the .NET version currently used by PowerShell by running one of the following commands:
    * `[System.Reflection.Assembly]::GetExecutingAssembly().ImageRuntimeVersion`
    * `$PSVersionTable`

> 2. If the .NET version is different than 4, create the following configuration file:

Reviewer 2: Perhaps we could combine these two steps - I think we are talking to quite technically-savvy readers for this document:
```suggestion
2. If Powershell is not using .NET version 4, put the following configuration into `{path to powershell executable}\powershell.exe.config` (for example `C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe.config`):
```
followed by the code block.

Author: I think I'd prefer to keep it the way it is, not because of the audience (I agree they are highly technical) but because it makes the sentences shorter and easier to parse.

    * File location – the folder that contains the *powershell.exe* file, for example *C:\Windows\System32\WindowsPowerShell\v1.0*
    * File name: *powershell.exe.config*
3. In the *powershell.exe.config* file, provide the following configuration:
    
    ```text {linenos=table}
    <?xml version="1.0"?> 
    <configuration> 
    <startup useLegacyV2RuntimeActivationPolicy="true"> 
        <supportedRuntime version="{<Latest build of .NET version 4, for example, v4.0.30319>}"/> 
        <supportedRuntime version="{<.NET version currently used by PowerShell, for example, v2.0.50727>}"/> 
    </startup> 
    </configuration> 
    ```
With the above configuration, PowerShell primarily uses .NET version 4, but also supports the version that it previously used, for example, version 2.

> ### 5.2 Could Not Load File or Assembly System.Management.Automation

Reviewer 2: This error message is so similar to the previous one, I wonder if we could add something to 5.1 to make it clear that we haven't accidentally duplicated the same error?

There is also a chance that the customer is getting the error in 5.2, but uses the steps in 5.1 because they think that is the error they have received.

Don't worry if you can't - but that was my first thought on reading this.

Author: I think it's worth looking into. @jveldhuizen @tieniber is the error in 5.2 related to 5.1 or are there completely different and just happen to generate a similar error message?

Reviewer 3: I never have had these errors, I think, but they are really different. The first one is about an incorrect .NET version, the second one is about one missing library on your computer. Because the solutions are really different, merging them to one will make the problems even more difficult to understand or to distinguish. For me personally, the 'cause' parts make it very clear what the problems are, but maybe that's different for people not so used to .NET.

Author: Alright, thank you! I'll keep them as is, I imagine whoever reads this is likely to be familiar with .NET.

---
title: "Automate Mendix Deployment on Microsoft Windows"
url: /developerportal/deploy/automate-mendix-deployment-on-microsoft-windows/
description: "How to automate Mendix deployment on servers running Windows"
weight: 5
tags: ["deploy", "Windows", "Mendix Service Console", "Microsoft", "CI/CD pipeline", "cmdlets"]
---

## 1 Introduction

> On Windows servers, instead of deploying your application manually, you can automate that part of your CI/CD pipeline by using [Power Shell cmdlets](https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/cmdlet-overview). Automating deployment makes the process of updating your application faster and lessens the potential for user errors by replacing the manual deployment steps with automation scripts. This document describes the required configuration, and provides sample automation scripts that you can use as basis for scripting your own deployment.

Reviewer 3: I'm not very sure, but I think Microsoft never uses a space between Power and Shell:

```suggestion
On Windows servers, instead of deploying your application manually, you can automate that part of your CI/CD pipeline by using [PowerShell cmdlets](https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/cmdlet-overview). Automating deployment makes the process of updating your application faster and lessens the potential for user errors by replacing the manual deployment steps with automation scripts. This document describes the required configuration, and provides sample automation scripts that you can use as basis for scripting your own deployment.
```

## 2 Prerequisites

Before starting this how-to, make sure you have completed the following prerequisites:

> * Manually deploy your Mendix app and ensure that there are no errors during the deployment. For more information, see [Microsoft Windows](/developerportal/deploy/deploy-mendix-on-microsoft-windows/).

Reviewer 3: It's at least important that the database exist. The `SynchronizeDatabase` parameter will not create the database for you, as the UI can do for you.

Author: @jveldhuizen The document I linked already mentions creating the db, do you think we need to point it out separately?

Reviewer 3: No indeed, it's fine so.

Author: Great, thank you for confirming!

* Ensure that your Windows Service Console version is 4.3 or above.

> * Familiarize yourself with the update process for Mendix apps running on Windows. For more information, see [MS Windows: Update a Mendix App](/developerportal/deploy/updating-a-mendix-application/).

Reviewer 3: A requirement at this moment is to use Windows PowerShell 5.1. You can say that we only support that version and not PowerShell (Core) 6 or higher. At this moment, the `Start-MxApp` command is giving an error when you run it on PowerShell 7.2.6, due to a different .NET version. We will support that in the future.
Be aware of the fact that Microsoft uses the name Windows PowerShell until version 5.1, PowerShell Core for version 6/6.1/6.2 and PowerShell for version 7/7.1/7.2.  
Instead of adding this requirement to the prerequisites, you can put this information in the Troubleshooting part:

The following exception
`Start-MxApp: Could not load file or assembly 'Mendix.Service, Version=4.7.0.0, Culture=neutral, PublicKeyToken=null'. Het systeem kan het opgegeven bestand niet vinden.`
can mean that you are not running the `Start-MxApp` cmdlet in Windows PowerShell 5.1, but in a higher version of PowerShell.

Or you do both, what you think is the most clear.

Author: Added both, let me know if you have any corrections.

2. In Windows PowerShell, run the following command: `Import-Module '{<Mendix Service Console installation directory>}\Mendix.Service.Commands.dll'`.
    For example, if Mendix Service Console is installed at *C:\Program Files (x86)\Mendix\Service Console*, enter `Import-Module 'C:\Program Files (x86)\Mendix\Service Console\Mendix.Service.Commands.dll'`
3. Verify that the following commands are now available in PowerShell:
    * `Start-MxApp`
    * `Stop-MxApp`
    * `Update-MxApp`
    * `Install-MxServer`
    * `Get-MxApp`
    
    For more information about each command and its parameters, add `-?` after the command, for example, `Start-MxApp -?`.

After installing the Mendix-specific cmdlets, you can use them to write your own scripts that start, stop, or update your Mendix app.

## 4 Sample Scripts

In this section, you can find sample scripts to help you script your app deployment.

{{% alert color="info" %}}
The scripts are intended to show the range of available deployment options. They are presented as examples only, and may require significant adaptation to work in your own environment.
{{% /alert %}}

### 4.1 Sample Script - Update App

> The following script example demonstrates the process required to update your app. First, it imports the necessary cmdlets. After that, it stops your app and updates it with files extracted from the Mendix Deployment Package. Finally, the script restarts the app with the `SynchronizeDatabase` parameter to synchronize the database without user input. Without this parameter, you must give additional confirmation to execute the database commands.

Reviewer 3: > Without this parameter, you must give additional confirmation to execute the database commands.

This is not true. Without this parameter, you will get this error:
```
Start-MxApp : Unable to start the app. Reason: The database is not synchronized with the model.
At line:1 char:1
+ Start-MxApp MyFirstApp
+ ~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (MyFirstApp:String) [Start-MxApp], Exception
    + FullyQualifiedErrorId : AppProcessError,Mendix.Service.Commands.StartAppCommand
```

Author: Removed the sentence. @jveldhuizen do you think this is worth mentioning in the troubleshooting section? How would they solve this? Just add the parameter and re-run the script?

Reviewer 3: Yes, it's fixed when they add that parameter and run it again.
I don't know what we should add to the Troubleshooting section. Is it only for the really exceptional and unexpected cases and weird errors, or also for these kind of problems people can have when they forget something? I don't have a strong opinion.
Preferable, we have a section where we describe each Cmdlet command more in detail. Now, you only explain this parameter in a sample script. There 's no section where we describe `Start-MxApp` itself. However, that's not the goal of this How-to and secondly, you already describe how to get more info about the commands, with `-?`. I think I should spend a littlebit more time in extending that built-in documentation. Even in this exception, I could mention the `SynchronizeDatabase` parameter, so that it's immediately clear. So that's on my todo list.

Author: I would say that for the troubleshooting, we should add whatever the client is likely to encounter. If an error is likely to happen, or if someone on our side has already encountered it, then we should document it, even if it's caused by user error. On the other hand, if the error is truly specific to our test setup and is unlikely that anyone else would have the same conditions necessary for the error to happen, then I'd say we don't need to document it.

I'm happy to add more info about the commands. I can write a basic description of them based on the scripts and my own best guess, but if there's a document I can reference, that would also be very helpful. I don't necessarily want to list all available parameters though, as I'm guessing those might change more often than the commands themselves.

On the other hand, the basic function of the commands is self-explanatory for the most part, so I'd be writing something like, "StartMxApp - starts the Mendix app", and that doesn't really add a lot of value. What do you think? Is there something else I could add without duplicating the results of the -? command?

Author: Forgot to add - I'm going to try and add a troubleshooting section for the above error too.

Author: @jveldhuizen I've added 1 more troubleshooting section per the above comment. Let me know if there's anything you'd like me to add, otherwise I plan to merge the PR and make the article available tomorrow.

```text {linenos=table}
Import-Module '{<Mendix Service Console installation directory>}\Mendix.Service.Commands.dll'                                                                                                                        
$MDA_PATH = Read-Host '{Location of your Mendix Deployment Package}'       
$APP_NAME = Read-Host '{Name of your app}'         

"Deploying " + $MDA_PATH + " to app " + $APP_NAME                         

# stop app                                                      
Stop-MxApp $APP_NAME                                                    

# unpack app                                                    
Update-MxApp $APP_NAME -LiteralPath $MDA_PATH + "/" + $MDA_FILE

# start app, update database                                     

> Start-MxApp MyFirstApp -SynchronizeDatabase

Reviewer 3: ```suggestion
Start-MxApp $APP_NAME -SynchronizeDatabase
```

### 5.2 Could Not Load File or Assembly System.Management.Automation

PowerShell shows an error message similar to the following:

```text {linenos=false}
Could not load file or assembly 'System.Management.Automation, Version=3.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
```

#### 5.2.1 Cause

Windows Management Framework 3.0 is not installed.

#### 5.2.2 Solution

Install [Windows Management Framework 3.0](https://www.microsoft.com/en-us/download/details.aspx?id=34595).

### 5.3 Cannot Open {App Name} Service on Computer

PowerShell shows an error message similar to the following:

```text {linenos=false}
Start-MxApp : Cannot open App1 service on computer '.'.
At line:1 char:1
+ Start-MxApp App1 -synchronizedatabase
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (App1:String) [Start-MxApp], InvalidOperationException
    + FullyQualifiedErrorId : AppProcessError,Mendix.Service.Commands.StartAppCommand
```
#### 5.3.1 Cause

PowerShell is run without administrator privileges.

#### 5.3.2 Solution

> Run PowerShell as an administrator.

Reviewer 3: Maybe useful:

If you do `Start-MxApp MyFirstApp` or even `Start-MxApp MyFirstApp -synchronizedatabase` and you get the following error:

```
Start-MxApp : Unable to start the app. Reason: The database does not exist.
At line:1 char:1
+ Start-MxApp MyFirstApp -synchronizedatabase
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (MyFirstApp:String) [Start-MxApp], Exception
    + FullyQualifiedErrorId : AppProcessError,Mendix.Service.Commands.StartAppCommand
```

you first have to create the database manually. There is no automatic way to create the database by a Windows Service cmdlet.

Author: Added section, please let me know if anything needs correcting.

Reviewer 3: Can you put `Start-MxApp : Unable to start the app. Reason: The database does not exist.` on top of the exception block?

Author: Fixed.
